### PR TITLE
[core] Timeout object fetches that take too long

### LINKED ISSUE #19565: [Feature] Throw an exception when object fetch times out
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

We should throw an ObjectLostError when an object fetch times out because there are no locations found.

This is important for lineage reconstruction so that the user will receive an error if there is a bug in reconstruction.

### Use case

A task or actor requesting an object with no locations should receive an error if the object will never appear. Instead right now, this can hang forever.

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

Remerging #19789 with some fixes for Dask-on-Ray 1TB sort:
- Fixes a bug where the timer was not getting reset correctly
- Increased timeout to 10min just to be safe
- Changed the error to a unique exception `ObjectFetchTimedOutError` to improve debugging. This exception should usually indicate a system-level bug.

## Related issue number

Closes #19565.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/object_manager/test/pull_manager_test.cc [hunk 1]
```diff
@@ -1078,6 +1088,97 @@ TEST_F(PullManagerWithAdmissionControlTest, TestPrioritizeWorkerRequests) {
...
+  fake_time_ += 601;
+  pull_manager_.Tick();
+  ASSERT_TRUE(timed_out_objects_.count(oids[0]));
+  timed_out_objects_.clear();
+  RAY_UNUSED(pull_manager_.CancelPull(req_id));
+
+  AssertNoLeaks();
+}
+
+TEST_P(PullManagerTest, TestTimeOutAfterFailedPull) {
```
Comment (stephanie-wang) [inline src/ray/object_manager/test/pull_manager_test.cc:1140]: I'm not sure how to trigger the Python-level error without adding some sort of chaos testing to the gRPC/pubsub layer, but here is a unit test for the pull manager that captures the bug (I think).

File: src/ray/object_manager/pull_manager.cc [hunk 2]
```diff
@@ -481,9 +492,24 @@ void PullManager::TryToMakeObjectLocal(const ObjectID &object_id) {
...
+    RAY_LOG(WARNING) << "Object neither in memory nor external storage "
+                     << object_id.Hex();
+    request.expiration_time_seconds =
+        get_time_seconds_() +
+        RayConfig::instance().fetch_fail_timeout_milliseconds() / 1e3;
+  } else if (request.pending_object_creation) {
+    // Object is pending creation, wait for the task that creates the object to
+    // finish.
+    RAY_LOG(INFO) << "Object pending creation " << object_id.Hex();
+    request.expiration_time_seconds =
```
Comment (rkooo567) [inline src/ray/object_manager/pull_manager.cc:505]: Is this the main change?

**Review by rkooo567 — APPROVED** [review]

File: src/ray/object_manager/test/pull_manager_test.cc (hunk 1 above)
Comment (rkooo567) [reply src/ray/object_manager/test/pull_manager_test.cc:1140]: Can you create an issue to write a python test after we merge https://github.com/ray-project/ray/pull/19970?

Comment (rkooo567) [reply src/ray/object_manager/test/pull_manager_test.cc:1140]: I think (hopefully) that can give us good signals in the future from Ci.

File: src/ray/object_manager/pull_manager.cc (hunk 2 above)
Comment (stephanie-wang) [reply src/ray/object_manager/pull_manager.cc:505]: Ah no, actually this was here already. I'll comment to show you the new change.

File: src/ray/object_manager/pull_manager.cc [hunk 3]
```diff
@@ -568,6 +595,8 @@ void PullManager::UpdateRetryTimer(ObjectPullRequest &request,
 
   // Bound the retry time at 10 * 1024 seconds.
   request.num_retries = std::min(request.num_retries + 1, 10);
+  // Also reset the fetch expiration timer, since we just tried this pull.
+  request.expiration_time_seconds = 0;
```
Comment (stephanie-wang) [inline src/ray/object_manager/pull_manager.cc:599]: This was the change that I think should fix the bug. Basically, we need to reset the fetch expiration timer if we do a successful pull.

Comment (rkooo567) [reply src/ray/object_manager/pull_manager.cc:599]: ah this makes sense! Dask on Ray 1TB has a lot of inefficient pulling issues, so this was manifested here I guess...

Comment (rkooo567) [reply src/ray/object_manager/pull_manager.cc:599]: (it is like inefficient code path kind of helps us revealing issues lol)

**Comment by rkooo567** [comment]
Maybe we can run nightly tests 2~3 times and merge the PR?

**Comment by stephanie-wang** [comment]
Passed dask_on_ray_1tb_sort on 3/3 tries.

**Comment by stephanie-wang** [comment]
@ericl can you approve the .proto change?

**Review by ericl — APPROVED** [review]
Proto lgtm

**Comment by rkooo567** [comment]
Btw can we create an issue to add a test in ci once grpc delay is supported?

**Comment by stephanie-wang** [comment]
> Btw can we create an issue to add a test in ci once grpc delay is supported?

How about just adding this test to #19970? I don't think we should use issues to track these kinds of TODOs.

**Comment by rkooo567** [comment]
Oh yeah that works too (maybe you can push the test?) any way to make us not forget about is fine!

**Pull request merged**